man/rmcp_ping.1
.TH rmcp_ping 1 02/24/06 OpenIPMI "IPMI RMCP Ping program"

.SH NAME
rmcp_ping \- IPMI RMCP Ping program

.SH SYNOPSIS
.B rmcp_ping
.RB [ \-p
.IR port ]
.RB [ \-t
.IR waittime ]
.RB [ \-s
.IR starttag ]
.RB [ \-d ]
.I destination

.SH DESCRIPTION
The
.B rmcp_ping
program sends an RMCP ping packet onces a second to the destination,
printing unique responses it receives.

.SH OPTIONS
.TP
.BI \-p\  port
Set the destination \fIport\fP.  This is the standard RMCP port by default,
623.
.TP
.BI \-t\  waittime
Sets the number of seconds to wait for a response.  Ten by default.
.TP
.BI \-s\  starttag
There is an 8-bit tag value in every ping packet (0-254, 255 is
reserved), this program will cycle through the values starting at
starttag.  This is zero by default
.TP
.B \-d
Turns on debugging to standard output.
.TP
.I destination
The target address, default is the boradcast address (default 255.255.255.255)

.SH AUTHOR
.PP
Corey Minyard <cminyard@mvista.org>


ChangeLog
2007-10-02 Corey Minyard <cminyard@mvista.com>

	* include/OpenIPMI/internal/ipmi_fru.h, lib/normal_fru.c: Fix
	problems compiling with older compilers.
	* lib/oem_atca_fru.c: ditto

	* lib/ipmi_lan.c: Fixed a problem where the LAN code would pick
	rmcp+ defaults that might not be available.  Instead, pick the
	most secure available default.

	* lib/domain.c: Free mc_oem_handlers when shutting down.

	* */Makefile.am: Removed -R and -rpath from everything, per
	a bug report from Debian.

	* swig/python/Makefile.am, swig/python/openipmigui/Makefile.am:
	Use INSTALL_DATA, not INSTALL_PROGRAM, to install python code.

	* man/rmcp_ping.1: Some minor manpage fixups.

	* configure.in: Move to version 2.0.12.

2007-09-15 Corey Minyard <cminyard@mvista.com>

	* ui/ui_os.c: This file shouldn't use the ipmi_mem_alloc and
	ipmi_mem_free calls, as that introduces a circular dependency.
	Only use malloc and free here.

2007-09-12 Corey Minyard <cminyard@mvista.com>

	* cmdlang/cmd_mc.c: Fix an unsigned problem pointed out by Phil
	Estes.

2007-09-11 Corey Minyard <cminyard@mvista.com>

	* lib/ipmi_lan.c, include/OpenIPMI/ipmi_lan.h: Add support for a
	parameter to set the maximum number of outstanding messages on a
	LAN connection.  This used to be fixed to 2, but now can range
	from 1 to 63.

	* swig/python/openipmigui/gui_domainDialog.py: Remove some unused
	definitions.  Add support for "int" types of parameters, like the
	new maximum outstanding message count.  Fix a bug if the parm type
	is not supported.

	* lanserv/emu.c: Enable support for the reserve SDR commands.  Not
	having this confuses ipmitool.

2007-09-06 Corey Minyard <cminyard@mvista.com>

	* lib/sensor.c, lib/entity.c: Don't abort SDR handling if the
	device string is invalid.  Just go on and set the id to
	"**INVALID**"

2007-08-17 Corey Minyard <cminyard@mvista.com>

	* sample/dump_sensors.c, sample/Makefile.am: Added a little
	program to just dump all the sensors.

2007-07-23 Corey Minyard <cminyard@mvista.com>

	* lib/conn.c: Fix a possible race if an OEM check fails.

2007-07-11 Corey Minyard <cminyard@mvista.com>

	* lib/domain.c, include/OpenIPMI/internal/ipmi_oem.h:
	Added infrastructure for fixups for main SDRs.

	* lib/oem_intel.c: Add main SDR fixups for Intel SE7520
	motherboards.

2007-05-25 Corey Minyard <cminyard@mvista.com>

	* lib/mc.c: Fix the domain fully up get/put operations to
	work properly.  Before, the fully up could come early in
	some situations.

2007-05-24 Corey Minyard <cminyard@mvista.com>

	* ui/ui.c: Move the call to ipmi_init() to before where things
	may be allocated.  Thanks to Bruce Rosenkrantz for pointing
	this out.

2007-05-22 Corey Minyard <cminyard@mvista.com>

	* swig/python/openipmigui/gui_SoL.py: Add dumping useful
	information if data processing for terminal input raises an
	exception.

	* swig/python/openipmigui/_term.py: Add some missed range checking
	to make sure Y values are within range with vt100 H and f
	commands.

	* swig/python/OpenIPMI_lang.i: If a method raises an exception,
	print the method name.
	
2006-07-21 Gerhard Obrecht <varnhalt2 at users.sourceforge.net>

        * lib/oem_kontron_conn.c: Add more board support.
		
2007-03-05 Corey Minyard <cminyard@mvista.com>

	* lib/sensor.c, include/OpenIPMI/internal/ipmi_sensor.h: Added a
	new function to get/set whether the sensor should be used for
	presence.  This is a replacement for the ignore_if_no_entity,
	whose use is vague, and abstracts this operation a little better.
	* lib/control.c, include/OpenIPMI/internal/ipmi_control.h: ditto
	for controls.

	* lib/entity.c, lib/oem_atca.c, lib/oem_motorola_mxp.c: Switch
	over to the new presence ignore functions.
	* lib/oem_test.c, sample/example_oem.c: ditto

	* lib/entity.c: Other presence handling fixes.

	* cmdlang/Makefile.am: Remove the -L$(DESTDIR)$libdir, it was
	broken to add it.
	* glib/Makefile.am: ditto
	* lanserv/Makefile.am: ditto
	* lib/Makefile.am: ditto
	* swig/perl/Makefile.am: ditto
	* swig/python/Makefile.am: ditto
	* tcl/Makefile.am: ditto
	* ui/Makefile.am: ditto
	* unix/Makefile.am: ditto

	* lib/domain.c, lib/entity.c: Found some minor bugs pointed out
	by new compilers.

	* swig/python/Makefile.am: Create lib-dynload before installing
	into it.

2007-03-05 David Greigo <dgriego@mvista.com>

	* sample/ipmicmd.c: Fix problems with handling errors in
	non-interactive mode.

2007-02-25 Corey Minyard <cminyard@mvista.com>

	* configure.in, OpenIPMI.spec.in: Improve the handling of perl
	installation.  Also improve the tcl and glib12 handling to make it
	more automatic.

	* configure.in: Move to version 2.0.11

2007-02-09 Corey Minyard <cminyard@mvista.com>

	* tcl/Makefile.am: Add $(TCL_LIBS) to the test programs ld flags.

2007-01-10 Michael Chu <michaelychu@yahoo.com>

	* sdr.c: Fix ipmi_sdr_save related issues: use standard fetch size,
	correct offset and record id values, parameter checking and passing,
	do sdr reservation and check status after sdr clear.

2006-12-28 Corey Minyard <cminyard@mvista.com>

	* swig/python/openipmigui/gui_cmdwin.py: Fix a problem

2006-12-06 Corey Minyard <cminyard@mvista.com>

	* lanserv/Makefile.am: Added the utils library to the
	applications.

2006-12-04 Corey Minyard <cminyard@mvista.com>

	* swig/python/openipmigui/gui.py: Made more reasonable defaults
	for the initial window settings.

	* doc/IPMI.ltx, doc/SimpleMonitoring.dia, doc/ipmi.bib: Add some
	more docs and some use cases.

2006-11-30 Corey Minyard <cminyard@mvista.com>

	* lib/sensor.c, include/OpenIPMI/ipmiif.h.in: Some sensors are not
	readable, specifically event-only sensors and sensors with system
	software IDs as the owner.  Add a readable field to the sensor
	object and check on read operations to make sure it is readable.
	Technically, for SDRs with a system software ID owner, it might be
	better to make sure that threadhold and event enables are not
	done, but hopefully the SDRs are right.  Yeah, right, I know.
	

sample/rmcp_ping.c
/*
 * rmcp_ping.c
 *
 * OpenIPMI RMCP pinger
 *
 * Author: Montavista Software
 *         Corey Minyard <cminyard@mvista.com>
 *
 *  This program is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public License
 *  as published by the Free Software Foundation; either version 2 of
 *  the License, or (at your option) any later version.
 *
 *
 *  THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESS OR IMPLIED
 *  WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 *  MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 *  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
 *  OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 *  ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR
 *  TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 *  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 *  You should have received a copy of the GNU Lesser General Public
 *  License along with this program; if not, write to the Free
 *  Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <sys/time.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <sys/select.h>
#include <netdb.h>

unsigned char ping_msg[12] =
{
    0x06, /* RMCP version 1.0 */
    0x00, /* reserved */
    0xff, /* RMCP seq num, not used for IPMI */
    0x06, /* ASF message */
    0x00, 0x00, 0x11, 0xbe, /* ASF IANA */
    0x80, /* Presence Ping */
    0x00, /* Message tag */
    0x00, /* Reserved */
    0x00  /* Data Length */
};

static char *progname;

static void
usage(void)
{
    fprintf(stderr,
	    "Send an RMCP ping packet onces a second to the destination,\n");
    fprintf(stderr,
	    "printing unique responses it receives\n");
    fprintf(stderr,
	    "Usage:\n");
    fprintf(stderr,
	    "  %s [-p <port>] [-t <waittime>] [-s <starttag>]"
	    " [-d] [destination]\n",
	    progname);
    fprintf(stderr,
	    "    -p - Destination port, defaults to 623\n");
    fprintf(stderr,
	    "    -t - Sets the number of seconds to wait for responses"
	    " (default 10)\n");
    fprintf(stderr,
	    "    -s - Sets the start tag to start sending at"
	    " (0-254, default 0\n");
    fprintf(stderr,
	    "    -d - enable debugging\n");
    fprintf(stderr,
	    "    destination - the target address, default is the broadcast\n");
    fprintf(stderr,
	    "        address (default 255.255.255.255)\n");
    exit(1);
}

unsigned int port = 0x26f;
unsigned int waittime = 10;
unsigned int starttag = 0;
int debug_packet = 0;

struct socklist
{
    struct in_addr addr;
    struct socklist *next;
};
struct socklist *socklist = NULL;

static int
add_host(struct sockaddr *addr)
{
    struct socklist *curr = socklist;
    struct sockaddr_in *ip;

    if (addr->sa_family != AF_INET)
	return 0;

    ip = (struct sockaddr_in *) addr;
    while (curr) {
	if (ip->sin_addr.s_addr == curr->addr.s_addr)
	    break;
	curr = curr->next;
    }
    if (curr)
	return 0;
    curr = malloc(sizeof(*curr));
    curr->addr = ip->sin_addr;
    curr->next = socklist;
    socklist = curr;
    return 1;
}

int
main(int argc, char *argv[])
{
    unsigned char      rsp[28];
    int                sock;
    int                rv;
    struct sockaddr    dest;
    size_t             destlen;
    struct sockaddr    src;
    int                val;
    socklen_t          fromlen;
    fd_set             readfds;
    struct timeval     currtime;
    struct timeval     endtime;
    struct timeval     twait;
    char               host[200];
    char               serv[200];
    char               *addrname;
    int                send_next;
    int                i;


    progname = argv[0];

    for (i=1; i<argc; i++) {
	if (argv[i][0] != '-')
	    break;
	if (argv[i][1] == '\0')
	    usage();
	if (strcmp(argv[i], "--") == 0) {
	    i++;
	    break;
	} else if (strcmp(argv[i], "-p") == 0) {
	    i++;
	    if (i >= argc) {
		fprintf(stderr, "No parameter given for -p\n\n");
		usage();
	    }
	    port = strtoul(argv[i], NULL, 0);
	} else if (strcmp(argv[i], "-t") == 0) {
	    i++;
	    if (i >= argc) {
		fprintf(stderr, "No parameter given for -t\n\n");
		usage();
	    }
	    waittime = strtoul(argv[i], NULL, 0);
	} else if (strcmp(argv[i], "-s") == 0) {
	    i++;
	    if (i >= argc) {
		fprintf(stderr, "No parameter given for -s\n\n");
		usage();
	    }
	    starttag = strtoul(argv[i], NULL, 0);
	    if (starttag > 254) {
		fprintf(stderr, "Invalid start tag for -s\n\n");
		usage();
	    }
	} else if (strcmp(argv[i], "-d") == 0) {
	    debug_packet++;
	}
    }
    
    if (i < argc) {
	struct addrinfo hints, *res0;
	char ports[16];

	snprintf(ports, sizeof(ports), "%d", port);
	memset(&hints, 0, sizeof(hints));
	hints.ai_family = PF_UNSPEC;
        hints.ai_socktype = SOCK_DGRAM;
	rv = getaddrinfo(argv[i], ports, &hints, &res0);
	if (rv != 0) {
	    fprintf(stderr, "Unable to handle given address: %s\n\n",
		    gai_strerror(rv));
	    usage();
	}
	dest = *(res0->ai_addr);
	destlen = res0->ai_addrlen;
	freeaddrinfo(res0);
    } else {
	struct sockaddr_in *ip = (struct sockaddr_in *) &dest;
	ip->sin_family = AF_INET;
	ip->sin_port = htons(port);
	ip->sin_addr.s_addr = INADDR_BROADCAST;
	destlen = sizeof(*ip);
    }

    sock = socket(PF_INET, SOCK_DGRAM, IPPROTO_UDP);
    if (sock == -1) {
	perror("socket");
	exit(1);
    }

    /* We want it to be non-blocking. */
    rv = fcntl(sock, F_SETFL, O_NONBLOCK);
    if (rv) {
	close(sock);
	perror("fcntl(sock, F_SETFL, O_NONBLOCK)");
	exit(1);
    }

    val = 1;
    rv = setsockopt(sock, SOL_SOCKET, SO_BROADCAST, &val, sizeof(val));
    if (rv) {
	close(sock);
	perror("setsockopt(sock, SO_BROADCAST)");
	exit(1);
    }

    gettimeofday(&currtime, NULL);
    endtime = currtime;
    endtime.tv_sec += waittime;
    send_next = 1;
    for (;;) {
	if ((endtime.tv_sec < currtime.tv_sec)
	    || ((endtime.tv_sec == currtime.tv_sec)
		&& (endtime.tv_usec < currtime.tv_usec)))
	    break;

	if (send_next) {
	    ping_msg[9] = starttag;
	    starttag++;
	    if (starttag > 254)
		starttag = 0;
	    rv = sendto(sock, ping_msg, sizeof(ping_msg), 0, &dest, destlen);
	    if (rv < 0) {
		close(sock);
		perror("sendto");
		exit(1);
	    }
	    send_next = 0;
	}

	twait.tv_usec = 0;
	twait.tv_sec = 1;
	FD_ZERO(&readfds);
	FD_SET(sock, &readfds);
	rv = select(sock+1, &readfds, NULL, NULL, &twait);
	if (rv < 0) {
	    close(sock);
	    perror("select");
	    exit(1);
	}
	if (rv == 0) {
	    send_next = 1;
	    goto next;
	}
	
	fromlen = sizeof(src);
	rv = recvfrom(sock, rsp, sizeof(rsp), 0, &src, &fromlen);

	if (debug_packet && (rv > 0)) {
	    int i;
	    printf("Got packet:");
	    for (i=0; i<rv; i++) {
		if ((i % 16) == 0)
		    printf("\n   ");
		printf(" %2.2x", rsp[i]);
	    }
	    printf("\n");
	}

	if (rv < 0) {
	    perror("recvfrom");
	} else if (rv < 28) {
	    fprintf(stderr, "Invalid receive length: %d, should be 28\n", rv);
	} else if ((rsp[0] != 6) || (rsp[3] != 6) || (rsp[4] != 0x00)
		   || (rsp[5] != 0x00) || (rsp[6] != 0x11) || (rsp[7] != 0xbe)
		   || (rsp[8] != 0x40) || (rsp[11] < 16))
	{
	    fprintf(stderr, "Invalid ping response\n");
	} else {
	    if (! add_host(&src))
		goto next;
	    rv = getnameinfo(&src, fromlen, host, sizeof(host),
			     serv, sizeof(serv), 0);
	    if (rv) {
		struct sockaddr_in *ip = (struct sockaddr_in *) &src;
		addrname = inet_ntoa(ip->sin_addr);
	    } else
		addrname = host;
	    printf("%s", addrname);
	    if ((rsp[20] & 0x80) && ((rsp[20] & 0xf) == 0x01))
		printf(" IPMI");
	    printf("\n");
	}
    next:
	gettimeofday(&currtime, NULL);
    }

    return 0;
}
